### PR TITLE
Replace deprecated av_init_packet() calls

### PR DESCRIPTION
**Pull request by sfan5**
FFmpeg intends to remove `sizeof(AVPacket)` from its' public ABI and has hence deprecated usage of `av_init_packet`.
Explicit allocation functions (`av_packet_alloc` / `av_packet_free`) are available since ffmpeg 3.0.

Tested normal operation and encoding (w/o audio) under ASan, should work fine.

### CODE REVIEW

**Comment by sfan5** [comment]
That's older (bbbf3571ed) and will only go away when ffmpeg actually removes the member (since while it still exists, we _do_ need to set it 1).

File: audio/decode/ad_lavc.c [hunk 1]
```diff
@@ -168,10 +168,15 @@ static int send_packet(struct mp_filter *da, struct demux_packet *mpkt)
...
-    AVPacket pkt;
-    mp_set_av_packet(&pkt, mpkt, &priv->codec_timebase);
+    int ret;
+    if (mpkt) {
+        AVPacket *pkt = mp_alloc_av_packet(mpkt, &priv->codec_timebase);
+        ret = avcodec_send_packet(avctx, pkt);
+        mp_free_av_packet(pkt);
+    } else {
+        ret = avcodec_send_packet(avctx, NULL);
+    }
```
Comment (ericonr) [inline audio/decode/ad_lavc.c:178]: Could the packet be allocated in the initialization code and stored in `priv`? To avoid so many allocations.

**Review by ericonr — COMMENTED** [review]
These changes don't deal with the possibility of allocation failure, right? Since a bunch of these used to be stack allocations, that wasn't a concern...

Maybe a lot of allocations should be moved to initialization code, too? Both to avoid many repeated allocation/free and to have a place where it's easy/simple to fail.

**Comment by sfan5** [comment]
> These changes don't deal with the possibility of allocation failure, right?

Correct, I don't exactly know what mpv usually does (I think talloc aborts?) but the code changes base on the assumption that small allocations can't/shouldn't fail.

> Maybe a lot of allocations should be moved to initialization code, too?

While that would be strictly _better_ I'm not convinced if the extra effort is worth it.

**Comment by decent-name** [comment]
The libavformat change has the potential to break the mpv cache: https://github.com/mpv-player/mpv/blob/master/demux/cache.c#L246
So I reject the libavformat change in the first place. Just continue using on-stack avpacket.

**Comment by decent-name** [comment]
Looking at the ffmpeg headers, it seems it doesn't affect the side data stuff (for the cache) yet: http://git.videolan.org/?p=ffmpeg.git;a=blob;f=libavcodec/packet.h;#l305
They forgot to specify anything about it.
av_packet_init is deprecated though, so it needs to be replaced. The simplest way to do this would be calling av_packet_alloc once (like with pthread_once) to get the default AVPacket contents, store them in a global AVPacket variable, and free the allocated packet. Then copy it per assignment (basically memcpy) in all situations where av_init_packet was needed. Care has to be taken in cases where some fields were expected to be preserved (av_init_packet doesn't write all fields; I don't know if mpv code makes use of it). This would be a good way to evade ffmpeg bullshit without hammering malloc even harder.

**Comment by sfan5** [comment]
For future reference these were wm4's comments from 14th May:
<blockquote>
The libavformat change has the potential to break the mpv cache: https://github.com/mpv-player/mpv/blob/master/demux/cache.c#L246<br/>
So I reject the libavformat change in the first place. Just continue using on-stack avpacket.
</blockquote>

<blockquote>
Looking at the ffmpeg headers, it seems it doesn't affect the side data stuff (for the cache) yet: http://git.videolan.org/?p=ffmpeg.git;a=blob;f=libavcodec/packet.h;#l305<br/>
They forgot to specify anything about it.<br/>
av_packet_init is deprecated though, so it needs to be replaced. The simplest way to do this would be calling av_packet_alloc once (like with pthread_once) to get the default AVPacket contents, store them in a global AVPacket variable, and free the allocated packet. Then copy it per assignment (basically memcpy) in all situations where av_init_packet was needed. Care has to be taken in cases where some fields were expected to be preserved (av_init_packet doesn't write all fields; I don't know if mpv code makes use of it). This would be a good way to evade ffmpeg bullshit without hammering malloc even harder.
</blockquote>

**Pull request closed without merging**